crates/driver/src/tests/setup/blockchain.rs

```diff
@@ -688,55 +684,55 @@ async fn primary_account(web3: &DynWeb3) -> ethcontract::Account {
+    #[allow(dead_code)] // store `process` to later run its `Drop` handler
+    process: tokio::process::Child,
+    url: String,
 }
 
-impl Geth {
-    /// Setup a new geth instance.
-    async fn new() -> Self {
-        let http = reqwest::Client::new();
-        let res = http
-            .post(format!("http://localhost:{DEV_GETH_PORT}"))
-            .send()
-            .await
-            .unwrap();
-        let port = res.text().await.unwrap();
-        Self { port }
-    }
-
-    pub fn url(&self) -> String {
-        format!("http://localhost:{}", self.port)
+impl std::fmt::Debug for Node {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Node").field("url", &self.url).finish()
     }
 }
 
-// What we really want here is "AsyncDrop", which is an unsolved problem in the
-// async ecosystem. As a workaround we create a new runtime so that we can block
-// on the delete request. Spawning a task for this isn't enough because tokio
-// runtimes when they exit drop background tasks, like when a #[tokio::test]
-// function returns.
-impl Drop for Geth {
-    fn drop(&mut self) {
-        let port = std::mem::take(&mut self.port);
-        let task = async move {
-            let client = reqwest::Client::new();
-            client
-                .delete(&format!("http://localhost:{DEV_GETH_PORT}/{port}"))
-                .send()
-                .await
-                .unwrap();
+impl Node {
+    /// Spawn a new node instance.
+    async fn new() -> Self {
+        use tokio::io::AsyncBufReadExt as _;
+
+        let mut process = tokio::process::Command::new("anvil")
+            .arg("--port")
+            .arg("0") // use 0 to let `anvil` use any open port
+            .arg("--balance")
+            .arg("1000000")
+            .kill_on_drop(true)
+            .stdout(std::process::Stdio::piped())
+            .spawn()
+            .unwrap();
+
+        let parse_url_from_logs = async {
+            const NEEDLE: &str = "Listening on ";
+            let mut reader = tokio::io::BufReader::new(process.stdout.take().unwrap()).lines();
+            while let Some(line) = reader.next_line().await? {
+                if let Some(addr) = line.strip_prefix(NEEDLE) {
+                    return Ok(format!("http://{addr}"));
+                }
+            }
+            anyhow::bail!("could not find \"{NEEDLE}.*\" in anvil logs")
         };
-        let runtime = tokio::runtime::Builder::new_current_thread()
-            .enable_all()
-            .build()
+
+        let url = tokio::time::timeout(tokio::time::Duration::from_secs(1), parse_url_from_logs)
+            .await
+            .expect("finding anvil URL timed out")
             .unwrap();
-        // block_on must be called in a new thread because tokio forbids nesting
-        // runtimes.
-        let handle = std::thread::spawn(move || runtime.block_on(task));
-        handle.join().unwrap();
+        Self { process, url }
+    }
```

Comment:
I _think_ this will have issues. Specifically, I believe you need to spawn a background task that continues to read from stdout.

I'm not 100% sure about this, but I think either the `anvil` process will block on writing to stdout or some buffer will run out of space or some buffer will grow very big.

Reply:
So, I'm not 100% sure about this - so take it with a grain of salt. But it this uses something like [`pipe(3)`](https://linux.die.net/man/3/pipe) under the hood, then I would expect the thread writing to `stdout` to block.

It can be that it works because logs are sent to a separate thread for writing so it isn't blocking `anvil` (but logs aren't working as expected).

Reply:
I know we are using the `async` version of processes, but this is from the [Rust standard library](https://doc.rust-lang.org/std/process/index.html#handling-io), and I imagine both versions use similar machinery under the hood (they just read from the pipes file descriptors with different sync and async APIs):

> ```rust
> // If the child process fills its stdout buffer, it may end up
> // waiting until the parent reads the stdout, and not be able to
> // read stdin in the meantime, causing a deadlock.
> // Writing from another thread ensures that stdout is being read
> // at the same time, avoiding the problem.
> let mut stdin = child.stdin.take().expect("failed to get stdin");
> std::thread::spawn(move || {
>     stdin.write_all(b"test").expect("failed to write to stdin");
> });
> ```

Reply:
One suggestion here would be to setup a separate task that:
- reads stdout from the `anvil` process
- prints the line to the `driver`'s stdout (so that the anvil logs show up)
- uses a one-shot channel to send the port (from scanning the stdout)

Reply:
[567b13a](https://github.com/cowprotocol/services/pull/1830/commits/567b13a83b2740d4a2b1a797dab2ac2f613ea840) I decided to keep the error handling to a minimum and also only log `stdout` and not `stderr` as I don't expect anybody to read these logs. Also I made sure to not log them by default because they are very noisy compared to the already noisy regular logs.